Download grades directly Fix for #628 ---> Translations?

Paul-Geppert opened the pull request:
Fix for #628 
On students site, I changed the student_index.html, that only one button for downloading the grades is shown, if only one document with grade exists.
It would be great, for having translations for the description of the grade-documents (like Midterm Notes --> Zwischennote), but I didn't find a translation script for it. The version for downloading more than one grade file didn't use a translation for the description too, only for "Download Grades".

janno42 commented:
for translation see #623

Steditor commented on evap/student/templates/student_index.html:
```diff
@@ -117,4 +121,4 @@
         {% endif %}
     {% endfor %}
-{% endblock %}
+{% endblock %}
```
Please add a newline at the end of the file.

janno42 commented:
Looks good! Can you please do the same on the results page (`results/templates/results_course_detail.html`)?

And please squash all you commits into a single commit when you are done.

Paul-Geppert commented:
I finished it. It's possible to download the grades directly from student and course side now.
I will squash my commits, thanks for this. This commit should be the very last for this issue, if you don't count #623 .

janno42 commented on evap/results/templates/results_course_detail.html:
```diff
@@ -33,7 +33,11 @@
         <div class="panel-heading">
             <span class="panel-title">{% trans "Overview" %}</span>
-            {% if course.grade_documents.count > 0 and can_download_grades %}
+            {% if course.grade_documents.count = 1 and can_download_grades %}
```
==

janno42 commented on evap/student/templates/student_index.html:
```diff
@@ -87,7 +87,9 @@
                                         {% endif %}
                                     {% endif %}
-                                    {% if course.grade_documents.count > 0 and can_download_grades %}
+                                    {% if course.grade_documents.count = 1 and can_download_grades %}
```
`==`

karyon replied:
did you talk to @janno42 about this and the other similar comment or address this somewhere else?

Paul-Geppert replied:
No, we didn't talk about it. I tested both versions, and both work. If == instead of = is better, I could fix it.

janno42 replied:
yes, please use `==`

Paul-Geppert commented:
This commit should resolve #623 and #628. I removed the drop-down menu for downloading grades, if only one grade document is available.
To have a translation for grade documents, I added a description-field for german and english and removed the old field for description.
To reach that, we added migration files, which add the new attributes to the database, save the old data, set the unique-dependencies (AlterUniqueTogether) and remove the attribute 'description'.
Beware: It is not revertable! That comes from the fact, that the AlterUniqueTogether-operation needs existing column and data to work. When reverting, one has to fill the 'description'-column with data first, so the position of the AlterUniqueTogether-operation is wrong for reverting.
Would that be changed, the migration has it now is, wouldn't work.

karyon commented on evap/grades/forms.py:
```diff
@@ -26,11 +25,17 @@ def __init__(self, *args, **kwargs):
+
+    def clean_description_en(self):
+        description_en = self.cleaned_data.get('description_en')
+        if GradeDocument.objects.filter(course=self.instance.course, description_en=description_en).exclude(id=self.instance.id).exists():
```
while you're at it, could you make this filter ignore case so we can catch more user errors here? ctrl-f for `iexact` in the django queryset docs.

Paul-Geppert replied:
Okay, I will think about more details for the filter.

karyon commented on evap/grades/migrations/0009_translate_description.py:
```diff
@@ -0,0 +1,23 @@
+            model_name='gradedocument',
+            name='description',
+        ),
+    ]
```
i'll think about the migration tomorrow. just a quick thought, @janno42 is it possible to just rename description to description_de, create a new description_en and add a unique together on that one? (maybe make description a CharField beforehand)

janno42 replied:
yes, i think that should work. before changing the unique_together we still need the data migration that puts stuff into description_en, but then it should be possible.

karyon replied:
@Paul-Geppert do you want to try this? what i'd do is start over with the base commit, rename description_de, make it a CharField, create CharField description_en, create migration. then add the data migration, and then add the second unique_together and create another migration. do a commit, and then somehow replace the migrations in your second commit with the new ones.

if you want to do something else, let me know, then i'd do this. or we look at it together during the next hacking hours.

Paul-Geppert replied:
Yes, I will try it and show you the result next meeting.

karyon replied:
nice. let me know if you encounter any problems.

Paul-Geppert commented:
@karyon please take over the migrations. I did little bugfixes discussed last meeting like == instead of =

karyon commented:
added new migrations and merged it manually, thanks!
